src/harness/reference_models/pre_iap_filtering/fss_purge.py

```diff
@@ -0,0 +1,296 @@
+  return fss_entity
+
+def performPurge(cbsds, fss_entity):
+  """ Implementation of algorithm described in R2-SGN-29.
+  The list of grants to be purged based on the interference caused by their
+  out of band emission are identified and the grants are removed from the CBSDs.
+
+  Args:
+    cbsds: List of CbsdData objects that are in the neighborhood to an FSS.
+    fss_entity: The FssProtectionPoint tuple of the FSS for which the purge is to be performed.
+  """
+
+  while True:
+    # Calculate the OOBE interference value for each CBSD based on the grant with the highest frequency.
+    calculateOobeInterference(cbsds, fss_entity)
+    
+    # Sort the CBSDs based on the calculated interference value.
+    sorted_cbsds = sorted(cbsds, key=lambda cbsd: cbsd['oobe_interference'])
+    # Calculate the threshold value in dBm for the OOBE interference
+    oobe_threshold_value = -129 + 10 * (np.log10(REF_BW/interference.ONE_MHZ) - FSS_OOBE_MARGIN)
+    
+    # Find the largest index such that aggregation of the interferences do not cross 
+    # the OOBE threshold
+    oobe_interferences = [interference.dbToLinear(cbsd['oobe_interference']) for cbsd in sorted_cbsds]
+    cumulated_interference = np.cumsum(oobe_interferences)
+    index = np.searchsorted(cumulated_interference, interference.dbToLinear(oobe_threshold_value))
+    cbsds_to_purge = sorted_cbsds[index:]
+    sorted_cbsds = sorted_cbsds[:index]
+    # Remove the grants from the CBSDs that are in the purge list.
+    if cbsds_to_purge:
+      for index, cbsd in enumerate(cbsds_to_purge):
+        if cbsd['grants']:
+          # Get the grant that was used for the OOBE interference calculation, that
+          # is the grant with highest highFrequency.
+          grant_with_max_high_freq = getMaxFreqWithoutpurgeFlag(cbsd)
+    
+          # Get all the grants that share the same emission mask as the one that 
+          # was used for OOBE calculation
+          if (grant_with_max_high_freq['operationParam']['operationFrequencyRange']
+              ['highFrequency'] > 3690000000):
+
+            # Adding the -13 emission mask grants to the purge list
+            grants_to_purge_for_cbsd = [grant for grant in cbsd['grants']
+                                        if grant['operationParam']['operationFrequencyRange']
+                                        ['highFrequency'] > 3690000000]
+          else:
+            # Adding the -25 emission mask grants to the purge list
+            grants_to_purge_for_cbsd = [grant for grant in cbsd['grants']
+                                        if grant['operationParam']['operationFrequencyRange']
+                                        ['highFrequency'] <= 3690000000]
+
+          # Setting the purge_flag for the grants that need to be purged
+          for grant_items in grants_to_purge_for_cbsd:
+             for grant in cbsd['grants']:
+               if grant == grant_items:
+                 grant['purge_flag'] = True
```

Comment:
Does this line update the "cbsds" list with the purge_flag? I don't see any place where the cbsds list is updated in the while loop. All the operations are done based on the sorted_cbsds list which is a sorted copy of the cbsds list.

Reply:
Re-modified   the code logic